Add gradient pt parameterization from Zhang+23

wbalmer opened the pull request:
Hi Tomas! This is my attempt to add the dlnP/dlnT and associated default number of nodes/priors on those nodes as detailed in [ZJ Zhang's paper](https://ui.adsabs.harvard.edu/abs/2023AJ....166..198Z/abstract). I tested it briefly and it ran okay, but there's still room for some improvement (for instance, ZJ mentions that 6 layers is an arbitrary definition, and more or less could be retrieved).

Let me know what you think - I'm not sure if I missed anything that would make this break e.g. the plots...

tomasstolker commented:
Fantastic! Thanks a lot William 🙏!

Did you make any changes to the `dTdP_temperature_profile` function from `pRT`? If not, then I would suggest to import it from `pRT` directly since that package is anyway used.

However, instead of importing at the top of the file, please import it in the function where it is used, because importing `pRT` is, just like `species`, quite slow. So I only imported function from `pRT` when they are actually used.

When the function that contains the import is ran a second (or more) time then it will be fast though (from what I could tell) because it is already part of `sys.modules`.

wbalmer commented:
Hi Tomas! I understand. I initially copied the function over intending to make some changes, and also because I was worried that the pRT version 3 release might break the imports, but I think the function is fine as it is right now (and we'll probably have to overhaul the imports anyways when pRT v3 comes out). This latest commit includes what I think is a safe import of the function where it is called in the `create_pt_profile` function. 

I also made the multinest arguments `const_efficiency_mode`, `sampling_efficiency`, and `evidence_tolerance`, optional arguments in `run_multinest` as opposed to hard coded. Is this okay?

tomasstolker commented on species/fit/retrieval.py:
```diff
@@ -787,6 +796,9 @@ def run_multinest(
         fit_corr: Optional[List[str]] = None,
         cross_corr: Optional[List[str]] = None,
         n_live_points: int = 2000,
+        const_efficiency_mode: Optional[bool] = True,
```
Thanks, these parameters are useful to include indeed! Perhaps we could add a `multinest_kwargs` parameter? So dictionary which then can also includes `n_live_points` and `resume`?

tomasstolker replied:
Perhaps for now leave in the `resume` and `n_live_points` though. We can add a deprecation warning.

wbalmer replied:
I added `mutlinest_kwargs` and a deprecation warning for `resume` and `n_live_points`, but am not sure if it is implemented how you intended. Let me know!

tomasstolker replied:
This looks great, thanks!

tomasstolker commented on species/fit/retrieval.py:
```diff
@@ -1056,6 +1068,17 @@ def run_multinest(
         else:
             self.abund_nodes = None
 
+        # set default gradient priors if applicable
```
Could you run [black](https://black.readthedocs.io) on this file? It will automatically fix some style things.

wbalmer replied:
I did this via a web app in commit [7be8239](https://github.com/tomasstolker/species/pull/82/commits/7be8239fcefc3e54103de0656dd4a2638e1300f8), hopefully that worked?

tomasstolker replied:
Thanks, yes looks good!

tomasstolker commented on species/fit/retrieval.py:
```diff
@@ -1056,6 +1068,17 @@ def run_multinest(
+            if prior is None:
+                prior = {}
+            for i in range(1,7):
+                if "PTslope_"+str(i) not in prior.keys():
```
- I think it would also work without the `.keys()`
- The string could be change to an f-string: `f"PTslope_{i}"`

tomasstolker commented on species/fit/retrieval.py:
```diff
@@ -1508,6 +1531,31 @@ def prior_func(cube, n_dim: int, n_param: int) -> None:
+                    )
+                else:
+                    # Default: 2000 - 12000 K
+                    tbottom = 2000.0 + 10000.0 * cube[cube_index["T_bottom"]]
```
Maybe the lower bound could be a bit smaller? Like 500 or 1000 K?

tomasstolker commented on species/fit/retrieval.py:
```diff
@@ -1508,6 +1531,31 @@ def prior_func(cube, n_dim: int, n_param: int) -> None:
+
+                cube[cube_index["T_bottom"]] = tbottom
+
+                for i in range(1,7):
```
Could also use f-strings here

tomasstolker commented on species/read/read_radtrans.py:
```diff
@@ -633,6 +633,26 @@ def get_model(
+
+            try:
+                from petitRADTRANS.physics import dTdP_temperature_profile
+            except ImportError:
```
An error would occur in this case below when `dTdP_temperature_profile` is called? Instead (I think) you could perhaps use `raise` within the `except` to create and error with the message from `print`

wbalmer replied:
I think I did this except as e, raise statement right?

tomasstolker replied:
Yes I think what you did is correct. So when it goes to the `except` it will still raise an error

tomasstolker reviewed (changes requested):
Very nice! Added a few comments for some minor improvements

tomasstolker commented on species/util/retrieval_util.py:
```diff
@@ -487,6 +487,28 @@ def create_pt_profile(
+
+        try:
+            from petitRADTRANS.physics import dTdP_temperature_profile
+        except ImportError:
```
Add the `Raise` also here?

tomasstolker commented:
A few more suggestions:

- The new parameters can be given a label (used for the posterior plot) in the `update_labels` function of `util.plot_util`
- The parameters can also be added to the list of `pt_param` in `plot_posterior` (lin 322) so they are skipped with `inc_pt_param=False`
- We could also update `plot_pt_profile`, but that is probably more work, so fine to leave it out if you don't have time for that!

wbalmer commented:
Done! I've updated `update_labels`, `pt_param` in `plot_posterior`, and added the raise where applicable. I added the parameters to `plot_pt_profile` in an earlier commit and can confirm it will produce a plot; were there more changes to that function you were thinking of? I am happy to try to implement them :)

tomasstolker commented on species/plot/plot_retrieval.py:
```diff
@@ -296,6 +299,24 @@ def plot_pt_profile(
+            try:
+                from petitRADTRANS.physics import dTdP_temperature_profile
+            except ImportError:
+                print("""Can\'t import the dTdP profile function from petitRADTRANS,
```
Add the `Raise` also here? Sorry I had missed this previously. Thanks for already having updated `plot_pt_profile`

tomasstolker commented on species/util/retrieval_util.py:
```diff
@@ -487,6 +487,30 @@ def create_pt_profile(
+        try:
+            from petitRADTRANS.physics import dTdP_temperature_profile
+        except ImportError as e:
+                raise ImportError(
```
The indentation seems off here?

tomasstolker commented:
I should probably have merged your PR before making changes to the structure of `AtmosphericRetrieval`... I can resolve this merge conflict.

tomasstolker commented:
I have merged the `main` branch into the PR.

@wbalmer, would you mind checking the _Files changed_ once more to see if I didn't accidentally introduce a error somewhere?

tomasstolker commented:
I commented out the `multinest_kwargs` for now, because for the new structure we can pass the parameters perhaps directly. Something that we can decide later. Sorry for the extra work you did...

wbalmer reviewed (commented):
No problem Tomas, thank you for helping merge this! The `multinest_kwargs` was a bit clunky, I prefer passing each option directly to the revamped `run_multinest`

tomasstolker commented:
Excellent, thanks a lot for the implementation! I will merge the PR 👍
